## Homework 1b: Training Language Models

In this assignment, you will implement several different types of language models for text.  We'll start with n-gram models. Then we'll build some neural n-gram and LSTM language models. Our primary dataset for this project will be the WikiText2 language modeling dataset.  This dataset comes with some of the basic preprocessing done for us, such as tokenization and rare word filtering (using the `<unk>` token). Therefore, we can assume that all word types in the test set also appear at least once in the training set. We'll also use the `torchtext` library to help with some of the data preprocessing, such as converting tokens into id numbers. **Warning: Do not start this project the day before it is due!  Some parts require 20 minutes or more to run, so debugging and tuning can take a significant amount of time.**

We've implemented a unigram model here as a demonstration.

In [ ]:
# This block handles some basic setup and data loading.  
# You shouldn't need to edit this, but if you want to 
# import other standard python packages, that is fine.

# imports
from collections import defaultdict, Counter
import numpy as np
import math
import tqdm
import random
import pdb

import torch
from torch import nn
import torch.nn.functional as F
import torchtext.legacy as torchtext

# download and load the data
text_field = torchtext.data.Field()
datasets = torchtext.datasets.WikiText2.splits(root='.', text_field=text_field)
train_dataset, validation_dataset, test_dataset = datasets

text_field.build_vocab(train_dataset, validation_dataset, test_dataset)
vocab = text_field.vocab
vocab_size = len(vocab)

train_text = train_dataset.examples[0].text # a list of tokens (strings)
validation_text = validation_dataset.examples[0].text

print("Number of words in the vocabulary: {}".format(vocab_size))
print("Example text: {}".format(validation_text[:30]))
print("Example text: {}".format(validation_text[1]))

downloading wikitext-2-v1.zip


100%|██████████| 4.48M/4.48M [00:00<00:00, 6.63MB/s]


extracting
Number of words in the vocabulary: 33279
Example text: ['<eos>', '=', 'Homarus', 'gammarus', '=', '<eos>', '<eos>', 'Homarus', 'gammarus', ',', 'known', 'as', 'the', 'European', 'lobster', 'or', 'common', 'lobster', ',', 'is', 'a', 'species', 'of', '<unk>', 'lobster', 'from', 'the', 'eastern', 'Atlantic', 'Ocean']
Example text: =


In [ ]:
class UnigramModel:
    def __init__(self, train_text):
        self.counts = Counter(train_text)
        self.total_count = len(train_text)

    def probability(self, word):
        return self.counts[word] / self.total_count

    def next_word_probabilities(self, text_prefix):
        """Return a list of probabilities for each word in the vocabulary."""
        return [self.probability(word) for word in vocab.itos]

    def perplexity(self, full_text):
        """Return the perplexity of the model on a text as a float.
        
        full_text -- a list of string tokens
        """
        log_probabilities = []
        for word in full_text:
            # Note that the base of the log doesn't matter 
            # as long as the log and exp use the same base.
            log_probabilities.append(math.log(self.probability(word), 2))
        return 2 ** -np.mean(log_probabilities)

unigram_demonstration_model = UnigramModel(train_text)
print('unigram validation perplexity:', 
      unigram_demonstration_model.perplexity(validation_text))

def check_validity(model):
    """Performs several sanity checks on your model:
    1) That next_word_probabilities returns a valid distribution
    2) That perplexity matches a perplexity calculated from next_word_probabilities

    Although it is possible to calculate perplexity from next_word_probabilities, 
    it is still good to have a separate more efficient method that only computes 
    the probabilities of observed words.
    """

    log_probabilities = []
    for i in range(10):
        prefix = validation_text[:i]
        probs = model.next_word_probabilities(prefix)
        assert min(probs) >= 0, "Negative value in next_word_probabilities"
        assert max(probs) <= 1 + 1e-8, "Value larger than 1 in next_word_probabilities"
        assert abs(sum(probs)-1) < 1e-4, "next_word_probabilities do not sum to 1"

        word_id = vocab.stoi[validation_text[i]]
        selected_prob = probs[word_id]
        log_probabilities.append(math.log(selected_prob))

    perplexity = math.exp(-np.mean(log_probabilities))
    your_perplexity = model.perplexity(validation_text[:10])
    assert abs(perplexity-your_perplexity) < 0.1, "your perplexity does not " + \
    "match the one we calculated from `next_word_probabilities`,\n" + \
    "at least one of `perplexity` or `next_word_probabilities` is incorrect.\n" + \
    f"we calcuated {perplexity} from `next_word_probabilities`,\n" + \
    f"but your perplexity function returned {your_perplexity} (on a small sample)."


check_validity(unigram_demonstration_model)

unigram validation perplexity: 965.0860734119312


To generate from a language model, we can sample one word at a time conditioning on the words we have generated so far.

In [ ]:
# Note: the prefix tokens will be used by our trigram language model
def generate_text(model, n=20, prefix=('<eos>', '<eos>')):
    prefix = list(prefix)
    for _ in range(n):
        probs = model.next_word_probabilities(prefix)
        word = random.choices(vocab.itos, probs)[0]
        prefix.append(word)
    return ' '.join(prefix)

print(generate_text(unigram_demonstration_model))

<eos> <eos> distinguished with other measures , fortune river first and me drove as 's Journal – at The . Nix sail


In fact there are many strategies to get better-sounding samples, such as only sampling from the top-k words or sharpening the distribution with a temperature.  You can read more about sampling from a language model in this paper: https://arxiv.org/pdf/1904.09751.pdf.

You will need to submit some outputs from the models you implement for us to grade.  The following function will be used to generate the required output files.

In [ ]:
!wget https://cal-cs288.github.io/sp21/project_files/proj_1/eval_prefixes.txt
!wget https://cal-cs288.github.io/sp21/project_files/proj_1/eval_output_vocab.txt
!wget https://cal-cs288.github.io/sp21/project_files/proj_1/eval_prefixes_short.txt
!wget https://cal-cs288.github.io/sp21/project_files/proj_1/eval_output_vocab_short.txt

def save_truncated_distribution(model, filename, short=True):
    """Generate a file of truncated distributions.
    
    Probability distributions over the full vocabulary are large,
    so we will truncate the distribution to a smaller vocabulary.

    Please do not edit this function
    """
    vocab_name = 'eval_output_vocab'
    prefixes_name = 'eval_prefixes'

    if short: 
      vocab_name += '_short'
      prefixes_name += '_short'

    with open('{}.txt'.format(vocab_name), 'r') as eval_vocab_file:
        eval_vocab = [w.strip() for w in eval_vocab_file]
    eval_vocab_ids = [vocab.stoi[s] for s in eval_vocab]
    print(len(eval_vocab_ids))

    all_selected_probabilities = []
    with open('{}.txt'.format(prefixes_name), 'r') as eval_prefixes_file:
        lines = eval_prefixes_file.readlines()
        for line in tqdm.tqdm_notebook(lines, leave=False):
            prefix = line.strip().split(' ')
            probs = model.next_word_probabilities(prefix)

            selected_probs = np.array([probs[i] for i in eval_vocab_ids], dtype=np.float32)
            all_selected_probabilities.append(selected_probs)

    all_selected_probabilities = np.stack(all_selected_probabilities)
    np.save(filename, all_selected_probabilities)
    print('saved', filename)

--2022-02-08 21:43:54--  https://cal-cs288.github.io/sp21/project_files/proj_1/eval_prefixes.txt
Resolving cal-cs288.github.io (cal-cs288.github.io)... 185.199.108.153, 185.199.109.153, 185.199.110.153, ...
Connecting to cal-cs288.github.io (cal-cs288.github.io)|185.199.108.153|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 519055 (507K) [text/plain]
Saving to: ‘eval_prefixes.txt’

eval_prefixes.txt   100%[===================>] 506.89K  --.-KB/s    in 0.01s   

2022-02-08 21:43:54 (37.9 MB/s) - ‘eval_prefixes.txt’ saved [519055/519055]

--2022-02-08 21:43:54--  https://cal-cs288.github.io/sp21/project_files/proj_1/eval_output_vocab.txt
Resolving cal-cs288.github.io (cal-cs288.github.io)... 185.199.108.153, 185.199.109.153, 185.199.110.153, ...
Connecting to cal-cs288.github.io (cal-cs288.github.io)|185.199.108.153|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12497 (12K) [text/plain]
Saving to: ‘eval_output_vocab.txt’

eval_output_

In [ ]:
save_truncated_distribution(unigram_demonstration_model, 'unigram_demonstration_predictions.npy')

560


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:29: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/1000 [00:00<?, ?it/s]

saved unigram_demonstration_predictions.npy


### N-gram Model

Now it's time to implement an n-gram language model.

Because not every n-gram will have been observed in training, use add-alpha smoothing to make sure no output word has probability 0.

$$P(w_2|w_1)=\frac{C(w_1,w_2)+\alpha}{C(w_1)+\alpha\cdot|V|}$$

where $|V|$ is the vocab size and $C$ is the count for the given bigram.  An alpha value around `3e-3`  should work.  Later, we'll replace this smoothing with model backoff. One edge case you will need to handle is at the beginning of the text where you don't have `n-1` prior words.  You can handle this however you like as long as you produce a valid probability distribution, but just using a uniform distribution over the vocabulary is reasonable for the purposes of this project.

A properly implemented bi-gram model should get a perplexity below 510 on the validation set. Please note that the trigram model will probably have a very high perplexity at this point, due to sparsity issues. We'll correct this below.

**Note**: Do not change the signature of the `next_word_probabilities` and `perplexity` functions.  We will use these as a common interface for all of the different model types.  Make sure these two functions call `n_gram_probability`, because later we are going to override `n_gram_probability` in a subclass. 
Also, we suggest pre-computing and caching the counts $C$ when you initialize `NGramModel` for efficiency. 

In [ ]:
class NGramModel:
    def __init__(self, train_text, n=2, alpha=3e-3):
        # get counts and perform any other setup
        self.n = n
        self.smoothing = alpha

        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.counts_w=Counter(list(zip(*[train_text[i:] for i in range(n)])))
        self.counts_c=Counter(list(zip(*[train_text[i:] for i in range(n-1)])))
        self.total_counts=len(train_text)+(n-1)
        self.total=len(train_text)  
    
        # END SOLUTION

    def n_gram_probability(self, n_gram):
        """Return the probability of the last word in an n-gram.
        
        n_gram -- a list of string tokens
        returns the conditional probability of the last token given the rest.
        """
        assert len(n_gram) == self.n

        # YOUR CODE HERE

        # BEGIN SOLUTION
        
        num=self.counts_w[tuple(n_gram)]+self.smoothing
        den=self.counts_c[tuple(n_gram[:-1])]+(self.smoothing*(vocab_size))
        if (self.n==1):
          num=self.counts_w[tuple(n_gram)]
          den=self.total
        return(num/den)
        # END SOLUTION

    def next_word_probabilities(self, text_prefix):
        """Return a list of probabilities for each word in the vocabulary."""

        # YOUR CODE HERE
        # use your function n_gram_probability
        # vocab.itos contains a list of words to return probabilities for

        # BEGIN SOLUTION
        text_prefix=['<eos>']*(self.n)+text_prefix
        return[self.n_gram_probability(((text_prefix)+(list(word.split(" "))))[-self.n:]) for word in vocab.itos]
        # END SOLUTION

    def perplexity(self, full_text):
        """ full_text is a list of string tokens
        return perplexity as a float """

        # YOUR CODE HERE
        # use your function n_gram_probability
        # This method should differ a bit from the example unigram model because 
        # the first n-1 words of full_text must be handled as a special case.

        # BEGIN SOLUTION
        log_probabilities = []
        full_text=['<eos>']*(self.n-1)+full_text
        for n_gram in list(zip(*[full_text[i:] for i in range(self.n)])):
          log_probabilities.append(math.log(self.n_gram_probability(n_gram), 2))
        return 2 ** -np.mean(log_probabilities)
        # END SOLUTION

unigram_model = NGramModel(train_text, 1)
check_validity(unigram_model)
print('unigram validation perplexity:', unigram_model.perplexity(validation_text)) # this should be the almost the same as our unigram model perplexity above

bigram_model = NGramModel(train_text, n=2)
check_validity(bigram_model)
print('bigram validation perplexity:', bigram_model.perplexity(validation_text))

trigram_model = NGramModel(train_text, n=3)
check_validity(trigram_model)
print('trigram validation perplexity:', trigram_model.perplexity(validation_text)) # this won't do very well...

save_truncated_distribution(bigram_model, 'bigram_predictions.npy') # this might take a few minutes

unigram validation perplexity: 965.0860734119312
bigram validation perplexity: 504.40791028428026
trigram validation perplexity: 2965.423598653458
560


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:29: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/1000 [00:00<?, ?it/s]

saved bigram_predictions.npy


Please download `bigram_predictions.npy` once you finish this section so that you can submit it.

In the block below, please report your bigram validation perplexity.  (We will use this to help us calibrate our scoring on the test set.)

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Bigram validation perplexity: ***504.4***

We can also generate samples from the model to get an idea of how it is doing.

In [ ]:
print(generate_text(bigram_model))

<eos> <eos> = = Province in southern solos Adummatu Peshkin dignitaries Sir Wiseman precipitated 470 chorus უ exacting Battlecruiser aspect of me


We now free up some RAM, **it is important to run the cell below, otherwise you will likely run out of RAM in the Colab runtime.**

In [ ]:
# Free up some RAM. 
del bigram_model
del trigram_model

This basic model works okay for bigrams, but a better strategy (especially for higher-order models) is to use backoff.  Implement backoff with absolute discounting: 
$$
\begin{align}
P\left(w_i|w_{i-n+1}^{i-1}\right)&=\frac{max\left\{C(w_{i-n+1}^i)-\delta,0\right\}}{\sum_{w_i} C\left(w_{i-n+1}^i\right)} + \alpha\left(w_{i-n+1}^{i-1}\right) P\left(w_i|w_{i-n+2}^{i-1}\right) \\
\alpha\left(w_{i-n+1}^{i-1}\right)&=\frac{\delta N_{1+}\left(w_{i-n+1}^{i-1}\right)}{{\sum_{w_i} C\left(w_{i-n+1}^i\right)}}
\end{align}
$$
where $N_{1+}$ is the number of words that appear after the previous $n-1$ words (the number of times the max will select something other than 0 in the first equation).  If $\sum_{w_i} C(w_{i-n+1}^i)=0$, use the lower order model probability directly (the above equations would have a division by 0). We found a discount $\delta$ of 0.9 to work well based on validation performance.  A trigram model with this discount value should get a validation perplexity below 275.

In [ ]:
s=['this', 'is' ,'yenna','this' ,'is' ,'cat','this' ,'is' ,'beka','this','is','this', 'are' ,'yenna','this', 'is']
n=1
c=list(zip(*[s[i:] for i in range(n)]))
w=(list(zip(*[s[i:] for i in range(2)])))
print(c)
print(w)
#Counter([each for each in set(c) for every in set(w) if each[:]==every[:-1]])
cn={}
for each in set(w):
  if each[1] in cn:
    cn[each[1]]+=1
  else:
    cn[each[1]]=1
print(cn)

[('this',), ('is',), ('yenna',), ('this',), ('is',), ('cat',), ('this',), ('is',), ('beka',), ('this',), ('is',), ('this',), ('are',), ('yenna',), ('this',), ('is',)]
[('this', 'is'), ('is', 'yenna'), ('yenna', 'this'), ('this', 'is'), ('is', 'cat'), ('cat', 'this'), ('this', 'is'), ('is', 'beka'), ('beka', 'this'), ('this', 'is'), ('is', 'this'), ('this', 'are'), ('are', 'yenna'), ('yenna', 'this'), ('this', 'is')]
{'this': 4, 'cat': 1, 'is': 1, 'yenna': 2, 'are': 1, 'beka': 1}


In [ ]:
class DiscountBackoffModel(NGramModel):
    def __init__(self, train_text, lower_order_model, n=2, delta=0.9):
        super().__init__(train_text, n=n)
        self.lower_order_model = lower_order_model
        self.discount = delta

        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.w=list(zip(*[train_text[i:] for i in range(n)]))
        self.counts_w=Counter(self.w)
        self.counts_c=Counter(list(zip(*[train_text[i:] for i in range(n-1)])))
        self.total_counts=len(train_text)+(n-1)
        self.total=len(train_text)
        self.counts_n={}
        self.counts_v={}
        for each in self.counts_w.keys():
          if each[:-1] in self.counts_n:
            self.counts_n[each[:-1]]+=1
          else:
            self.counts_n[each[:-1]]=1
        for each in self.w:
          if each[:-1] in self.counts_v:
            self.counts_v[each[:-1]]+=1
          else:
            self.counts_v[each[:-1]]=1     
        # END SOLUTION

    def n_gram_probability(self, n_gram):
        assert len(n_gram) == self.n

        # YOUR CODE HERE
        # back off to the lower_order model with n'=n-1 using its n_gram_probability function

        # BEGIN SOLUTION

        num=max(self.counts_w[tuple(n_gram)]-self.discount,0)         
        if tuple(n_gram[:-1]) not in self.counts_v.keys():
          return self.lower_order_model.n_gram_probability(n_gram[1:])
        else:
          den=self.counts_v[tuple(n_gram[:-1])]
          alpha=self.discount*self.counts_n[tuple(n_gram[:-1])]
          return (num/den) + (alpha/den) * self.lower_order_model.n_gram_probability(n_gram[1:])
      
        # END SOLUTION


bigram_backoff_model = DiscountBackoffModel(train_text, unigram_model, 2)
trigram_backoff_model = DiscountBackoffModel(train_text, bigram_backoff_model, 3)
check_validity(trigram_backoff_model)
print('trigram backoff validation perplexity:', trigram_backoff_model.perplexity(validation_text))

trigram backoff validation perplexity: 271.1028738950923


Free up RAM. 

In [ ]:
# Release models we don't need any more. 
del unigram_model
del bigram_backoff_model
del trigram_backoff_model

Now, implement Kneser-Ney to replace the unigram base model.
$$P(w)\propto |\{w':c(w',w) > 0\}|$$
A Kneser-Ney trigram model should get a validation perplexity below 260.

In [ ]:
class KneserNeyBaseModel(NGramModel):
    def __init__(self, train_text):
        super().__init__(train_text, n=1)

        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.counts_uni=Counter(list(zip(*[train_text[i:] for i in range(n)])))
        self.bigrams=list(zip(*[train_text[i:] for i in range(n+1)]))
        self.counts_bi=Counter(self.bigrams)
        self.counts_precede={}
        for each in self.counts_bi.keys():
          if each[1] in self.counts_precede:
            self.counts_precede[each[1]]+=1
          else:
            self.counts_precede[each[1]]=1
        
        # END SOLUTION

    def n_gram_probability(self, n_gram):
        assert len(n_gram) == 1

        # YOUR CODE HERE

        # BEGIN SOLUTION
        if n_gram[0] in self.counts_precede.keys():
          return self.counts_precede[n_gram[0]]/len(self.counts_bi.keys())
        else:
          return 0
        # END SOLUTION

kn_base = KneserNeyBaseModel(train_text)
check_validity(kn_base)
bigram_kn_backoff_model = DiscountBackoffModel(train_text, kn_base, 2)
trigram_kn_backoff_model = DiscountBackoffModel(train_text, bigram_kn_backoff_model, 3)
print('trigram Kneser-Ney backoff validation perplexity:', trigram_kn_backoff_model.perplexity(validation_text))

save_truncated_distribution(trigram_kn_backoff_model, 'trigram_kn_predictions.npy') # this might take a few minutes

trigram Kneser-Ney backoff validation perplexity: 256.5920656549996
560


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:29: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/1000 [00:00<?, ?it/s]

saved trigram_kn_predictions.npy


In [ ]:
print(generate_text(trigram_kn_backoff_model))
print(generate_text(trigram_kn_backoff_model, prefix=['What','about']))

<eos> <eos> <eos> " The Nightmare " had to work for expansion of the foreigners and administered scattered localities services at Cardiff
What about the CIA 's vocals span of Tanaka 's : by the European marketing policy and in the United States ,


Fill in your trigram backoff perplexities with and without Kneser Ney.

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Trigram backoff validation perplexity: ***271.102***

Trigram backoff with Kneser Ney perplexity: ***256.592***

Free up RAM. 

In [ ]:
# Delete models we don't need. 
del kn_base
del bigram_kn_backoff_model
del trigram_kn_backoff_model

If you want to learn more about n-gram language models and smoothing techniques, checkout the following paper: https://people.eecs.berkeley.edu/~klein/cs294-5/chen_goodman.pdf

### Neural N-gram Model

In this section, you will implement a neural version of an n-gram model.  The model will use a simple feedforward neural network that takes the previous `n-1` words and outputs a distribution over the next word.

You will use PyTorch to implement the model.  We've provided a little bit of code to help with the data loading using PyTorch's data loaders (https://pytorch.org/docs/stable/data.html)

A model with the following architecture and hyperparameters should reach a validation perplexity below 226.
* embed the words with dimension 128, then flatten into a single embedding for $n-1$ words (with size $(n-1)*128$)
* run 2 hidden layers with 1024 hidden units, then project down to size 128 before the final layer (ie. 4 layers total). 
* use weight tying for the embedding and final linear layer (this made a very large difference in our experiments); you can do this by creating the output layer with `nn.Linear`, then using `F.embedding` with the linear layer's `.weight` to embed the input
* rectified linear activation (ReLU) and dropout 0.1 after first 2 hidden layers. **Note: You will likely find a performance drop if you add a nonlinear activation function after the dimension reduction layer.**
* train for 10 epochs with the Adam optimizer (should take around 15-20 minutes)
* do early stopping based on validation set perplexity (see HW1A)


We encourage you to try other architectures and hyperparameters, and you will likely find some that work better than the ones listed above.  A proper implementation with these should be enough to receive full credit on the assignment, though.

In [ ]:
def ids(tokens):
    return [vocab.stoi[t] for t in tokens]

def append_eos(text_prefix):   
    prefix=["<eos>"]*(n-1-len(text_prefix))
    prefix.extend(text_prefix)
    return (prefix)

In [ ]:
assert torch.cuda.is_available(), "no GPU found, in Colab go to 'Edit->Notebook settings' and choose a GPU hardware accelerator"

class NeuralNgramDataset(torch.utils.data.Dataset):
    def __init__(self, text_token_ids, n):
        self.text_token_ids = text_token_ids
        self.n = n

    def __len__(self):
        return len(self.text_token_ids)

    def __getitem__(self, i):
        if i < self.n-1:
            prev_token_ids = [vocab.stoi['<eos>']] * (self.n-i-1) + self.text_token_ids[:i]
        else:
            prev_token_ids = self.text_token_ids[i-self.n+1:i]

        assert len(prev_token_ids) == self.n-1

        x = torch.tensor(prev_token_ids)
        y = torch.tensor(self.text_token_ids[i])
        return x, y

class NeuralNGramNetwork(nn.Module):
    # a PyTorch Module that holds the neural network for your model

    def __init__(self, n):
        super().__init__()
        self.n = n

        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.final=nn.Linear(128,vocab_size)
        self.embedding_layer=nn.Embedding(vocab_size,128,_weight=self.final.weight)
        self.flatten=nn.Flatten(1,2)
        self.linear_1=nn.Linear(128*(self.n-1),1024)
        self.relu=nn.ReLU()
        self.drop=nn.Dropout(p=0.1)
        self.linear_2=nn.Linear(1024,1024)
        self.linear_3=nn.Linear(1024,128)
                  
        
        # END SOLUTION

    def forward(self, x):
        # x is a tensor of inputs with shape (batch, n-1)
        # this function returns a tensor of log probabilities with shape (batch, vocab_size)

        # YOUR CODE HERE

        # BEGIN SOLUTION
        embedding=self.embedding_layer(x)
        out=self.flatten(embedding)
        out=self.linear_1(out)
        out=self.relu(out)
        out=self.drop(out)
        out=self.linear_2(out)
        out=self.relu(out)
        out=self.drop(out)
        out=self.linear_3(out)
        out=self.final(out)
        #self.embedding_layer.weight=self.final.weight

        return out
        # END SOLUTION

class NeuralNGramModel:
    # a class that wraps NeuralNGramNetwork to handle training and evaluation
    # it's ok if this doesn't work for unigram modeling
    def __init__(self, n):
        self.n = n
        self.network = NeuralNGramNetwork(n).cuda()

    def train(self,n_epochs=2):
        dataset = NeuralNgramDataset(ids(train_text), self.n)
        train = torch.utils.data.DataLoader(dataset, batch_size=128, shuffle=True)
        # iterating over train_loader with a for loop will return a 2-tuple of batched tensors
        # the first tensor will be previous token ids with size (batch, n-1),
        # and the second will be the current token id with size (batch, )
        # you will need to move these tensors to GPU, e.g. by using the Tensor.cuda() function.

        # this will take some time to run; use tqdm.tqdm_notebook to get a progress bar 
        # (see Project 0 for example)

        # YOUR CODE HERE

        # BEGIN SOLUTION
        adam=torch.optim.Adam(self.network.parameters())
        for epoch in range(n_epochs):
          print("Epoch: ",epoch)
          
          for batch in tqdm.tqdm_notebook(train, leave=False):
            words_batch, next_batch = batch
            adam.zero_grad()
            words_batch, next_batch = words_batch.cuda(), next_batch.cuda()
            out = self.network(words_batch)
            loss = F.cross_entropy(out, next_batch)
            loss.backward()
            adam.step()
          
          perplexity = self.perplexity(validation_text)
          print('Perplexity: ',perplexity)
          previous_best_score = 10000
          if perplexity < previous_best_score:
            print("Yay! New best score")
            previous_best_score = perplexity 
            torch.save(self.network.state_dict(),'neural_network.pt')

          self.network.load_state_dict(torch.load('neural_network.pt'))
        # END SOLUTION

    def next_word_probabilities(self, text_prefix):
        # YOUR CODE HERE
        # don't forget self.network.eval()
        # you will need to convert text_prefix from strings to numbers with the `ids` function
        # if your `perplexity` function below is based on a NeuralNgramDataset DataLoader, you will need to use the same strategy for prefixes with less than n-1 tokens to pass the validity check
        #   the data loader appends extra "<eos>" (end of sentence) tokens to the start of the input so there are always enough to run the network

        # BEGIN SOLUTION
        self.network.eval()
        text_prefix=['<eos>']*(self.n)+text_prefix
        
        n_gram=text_prefix[-(self.n-1):]
        n_gram=torch.LongTensor([ids(n_gram)]).cuda()
        logits=self.network(n_gram)
        softmax=F.softmax(logits,dim=1).squeeze(0)
        return(softmax.tolist())
        # END SOLUTION

    def perplexity(self, text):
        # you may want to use a DataLoader here with a NeuralNgramDataset
        # don't forget self.network.eval()

        # YOUR CODE HERE

        # BEGIN SOLUTION
        with torch.no_grad():
           self.network.eval()

           dataset = NeuralNgramDataset(ids(text),self.n)
           train = torch.utils.data.DataLoader(dataset, batch_size=128, shuffle=False)

           log = []

           for batch in tqdm.tqdm_notebook(train, leave=False):

             input, output = batch
             input, output = input.cuda(),output.cuda()
             logits = self.network(input)
             probabilities = F.softmax(logits, dim=1)
             out = torch.gather(probabilities, 1, output.unsqueeze(-1))

             for each in out:
               log.append(math.log(each.item(),2))

             
        return 2**-np.mean(log)
              

        # END SOLUTION


neural_trigram_model = NeuralNGramModel(3)
check_validity(neural_trigram_model)
neural_trigram_model.train(10)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:144: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/1 [00:00<?, ?it/s]

Epoch:  0


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:92: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Perplexity:  247.32009909485214
Epoch:  1


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Perplexity:  223.15805052511098
Epoch:  2


  0%|          | 0/16318 [00:00<?, ?it/s]

  0%|          | 0/1701 [00:00<?, ?it/s]

Perplexity:  226.4814841069515


In [ ]:
print('neural trigram validation perplexity:', neural_trigram_model.perplexity(validation_text))

save_truncated_distribution(neural_trigram_model, 'neural_trigram_predictions.npy', short=False)

Fill in your neural trigram perplexity.

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Neural trigram validation perplexity: ***223.15***

Free up RAM.

In [ ]:
# Delete model we don't need. 
del neural_trigram_model

### LSTM Model

For this stage of the project, you will implement an LSTM language model.

For recurrent language modeling, the data batching strategy is a bit different from what is used in some other tasks.  Sentences are concatenated together so that one sentence starts right after the other, and an unfinished sentence will be continued in the next batch.  We'll use the `torchtext` library to manage this batching for you.  To properly deal with this input format, you should save the last state of the LSTM from a batch to feed in as the first state of the next batch.  When you save state across different batches, you should call `.detach()` on the state tensors before the next batch to tell PyTorch not to backpropagate gradients through the state into the batch you have already finished (which will cause a runtime error).

We expect your model to reach a validation perplexity below 130.  The following architecture and hyperparameters should be sufficient to get there.
* 3 LSTM layers with 512 units
* dropout of 0.5 after each LSTM layer
* instead of projecting directly from the last LSTM output to the vocabulary size for softmax, project down to a smaller size first (e.g. 512->128->vocab_size). **NOTE: You may find that adding nonlinearities between these layers can hurt performance, try without first.**
* use the same weights for the embedding layer and the pre-softmax layer; dimension 128
* train with Adam (using default learning rates) for at least 20 epochs


In [ ]:
# This block handles some basic setup and data loading.  
# You shouldn't need to edit this, but if you want to 
# import other standard python packages, that is fine.

# imports
import os
os.environ['CUDA_LAUNCH_BLOCKING'] = "1"
from collections import defaultdict, Counter
import numpy as np
import math
import tqdm
import random
import pdb

import torch
from torch import nn
import torch.nn.functional as F
import torchtext.legacy as torchtext

# download and load the data
text_field = torchtext.data.Field()
datasets = torchtext.datasets.WikiText2.splits(root='.', text_field=text_field)
train_dataset, validation_dataset, test_dataset = datasets

text_field.build_vocab(train_dataset, validation_dataset, test_dataset)
vocab = text_field.vocab
vocab_size = len(vocab)

train_text = train_dataset.examples[0].text # a list of tokens (strings)
validation_text = validation_dataset.examples[0].text

print("Number of words in the vocabulary: {}".format(vocab_size))
print("Example text: {}".format(validation_text[:30]))
print("Example text: {}".format(validation_text[1]))

Number of words in the vocabulary: 33279
Example text: ['<eos>', '=', 'Homarus', 'gammarus', '=', '<eos>', '<eos>', 'Homarus', 'gammarus', ',', 'known', 'as', 'the', 'European', 'lobster', 'or', 'common', 'lobster', ',', 'is', 'a', 'species', 'of', '<unk>', 'lobster', 'from', 'the', 'eastern', 'Atlantic', 'Ocean']
Example text: =


In [ ]:
class LSTMNetwork(nn.Module):
    # a PyTorch Module that holds the neural network for your model

    def __init__(self):
        super().__init__()

        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.drop=nn.Dropout(p=0.5)
        self.final = nn.Linear(128,vocab_size)

        self.embedding=nn.Embedding(vocab_size,128)
        self.lstm = nn.LSTM(128,512,num_layers=3,dropout=0.5)
        self.linear = nn.Linear(512,128)
        # END SOLUTION

    def forward(self, x, state):
        """Compute the output of the network.
        
        Note: In the Pytorch LSTM tutorial, the state variable is named "hidden":
        https://pytorch.org/tutorials/beginner/nlp/sequence_models_tutorial.html

        The torch.nn.LSTM documentation is quite helpful:
        https://pytorch.org/docs/stable/nn.html#lstm
    
        x - a tensor of int64 inputs with shape (seq_len, batch)
        state - a tuple of two tensors with shape (num_layers, batch, hidden_size)
                representing the hidden state and cell state of the of the LSTM.
        returns a tuple with two elements:
          - a tensor of log probabilities with shape (seq_len, batch, vocab_size)
          - a state tuple returned by applying the LSTM.
        """

        # Note that the nn.LSTM module expects inputs with the sequence 
        # dimension before the batch by default.
        # In this case the dimensions are already in the right order, 
        # but watch out for this since sometimes people put the batch first

        # YOUR CODE HERE
        #seq,batch=x.size(dim=0),x.size(dim=1)
        #embeddings=self.embedding(x)
        #lstm_out,lstm_state=self.lstm(embeddings,state)
        #linear_out=self.linear(self.drop(lstm_out).flatten(0,1))
        #linear_out=self.linear(self.drop(lstm_out))
        #self.final.weight=self.embedding.weight
        #final_out=self.final(linear_out)
        #return (final_out.reshape(seq,batch,vocab_size),lstm_state)
        #return (final_out,lstm_state)
        # BEGIN SOLUTION
        embeddings=self.embedding(x)
        lstm_out,lstm_state=self.lstm(embeddings,state)
        linear_out=self.linear(self.drop(lstm_out))
        final_out=self.final(linear_out)
        self.embedding.weight = self.final.weight
        return (final_out,lstm_state)
        # END SOLUTION

class LSTMModel:
    "A class that wraps LSTMNetwork to handle training and evaluation."

    def __init__(self):
        self.network = LSTMNetwork().cuda()

    def train(self,n_epochs=5):
        train_iterator = torchtext.data.BPTTIterator(train_dataset, batch_size=64, 
                                                     bptt_len=32, device='cuda')
        # Iterate over train_iterator with a for loop to get batches
        # each batch object has a .text and .target attribute with
        # token id tensors for the input and output respectively.

        # The initial state passed into the LSTM should be set to zero.

        #dataset = NeuralNgramDataset(ids(train_text), self.n)
        # train = torch.utils.data.DataLoader(dataset, batch_size=128, shuffle=True)
        # iterating over train_loader with a for loop will return a 2-tuple of batched tensors
        # the first tensor will be previous token ids with size (batch, n-1),
        # and the second will be the current token id with size (batch, )
        # you will need to move these tensors to GPU, e.g. by using the Tensor.cuda() function.

        # this will take some time to run; use tqdm.tqdm_notebook to get a progress bar 
        # (see Project 0 for example)

        # YOUR CODE HERE

        # BEGIN SOLUTION
        adam=torch.optim.Adam(self.network.parameters())
        for epoch in range(n_epochs):
          print("Epoch: ",epoch)
          state = (torch.autograd.variable(torch.zeros(3,64,512)).cuda(),torch.autograd.variable(torch.zeros(3,64,512)).cuda())
          #print("state")
          for batch in tqdm.tqdm_notebook(train_iterator, leave=False):
            words_batch, next_batch = batch.text.cuda(), batch.target.cuda()
            adam.zero_grad()
            out,state = self.network(words_batch,(state[0].detach(),state[1].detach()))
            loss = F.cross_entropy(torch.permute(out,(1,2,0)).cuda(), next_batch.permute(1,0).cuda())
            loss.backward()
            adam.step()

          perplexity = self.dataset_perplexity(validation_dataset)
          print('Perplexity: ',perplexity)
          previous_best_score = 35000
          if perplexity < previous_best_score:
            previous_best_score = perplexity 
            torch.save(self.network.state_dict(),'network.pt')

          self.network.load_state_dict(torch.load('network.pt'))

        # END SOLUTION

    def next_word_probabilities(self, text_prefix):
        "Return a list of probabilities for each word in the vocabulary."

        prefix_token_tensor = torch.tensor(ids(text_prefix), device='cuda').view(-1, 1)
        
        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.network.eval()
        state = (torch.autograd.variable(torch.zeros(3,1,512)).cuda(),torch.autograd.variable(torch.zeros(3,1,512)).cuda())
        out,state=self.network(prefix_token_tensor,(state[0].detach(),state[1].detach()))
        softmax=(F.softmax(out,dim=2))[-1,:,:].squeeze()
        return(softmax.tolist())
        # END SOLUTION

    def dataset_perplexity(self, torchtext_dataset):
        "Return perplexity as a float."
        # Your code should be very similar to next_word_probabilities, but
        # run in a loop over batches. Use torch.no_grad() for extra speed.

        iterator = torchtext.data.BPTTIterator(torchtext_dataset, batch_size=64, bptt_len=32, device='cuda')

        # YOUR CODE HERE

        # BEGIN SOLUTION

        self.network.eval()
        log_probabilities=[]
        state = (torch.autograd.variable(torch.zeros(3,64,512)).cuda(),torch.autograd.variable(torch.zeros(3,64,512)).cuda())

        for batch in tqdm.tqdm_notebook(iterator, leave=False):
          with torch.no_grad():
            words_batch, next_batch = batch.text.cuda(),batch.target.cuda()
            out,state=self.network(words_batch,(state[0].detach(),state[1].detach()))
            softmax=F.softmax(out,dim=2)
            probs=torch.gather(softmax, 2, next_batch.unsqueeze(-1))
            for each in probs.squeeze(-1):
               for sent in each:
                log_probabilities.append(math.log(sent.item(),2))
        self.network.train()
        return 2**-np.mean(log_probabilities) 

        # END SOLUTION

lstm_model = LSTMModel()
lstm_model.train(20)

print('lstm validation perplexity:', lstm_model.dataset_perplexity(validation_dataset))
save_truncated_distribution(lstm_model, 'lstm_predictions_zero.npy', short=False)

<!-- Do not remove this comment, it is used by the autograder: RqYJKsoTS6 -->

Fill in your LSTM perplexity. 

LSTM validation perplexity: ***123.9797***

# Experimentation

Now it's time for you to experiment.  Try to reach a validation perplexity below 120. You may either modify the LSTM class above, or copy it down to the code cell below and modify it there. Just **be sure to run code cell below to generate results with your improved LSTM**.  

It is okay if the bulk of your improvements are due to hyperparameter tuning (such as changing number or sizes of layers), but implement at least one more substantial change to the model.  Here are some ideas (several of which come from https://arxiv.org/pdf/1708.02182.pdf):
* activation regularization - add a l2 regularization penalty on the activation of the LSTM output (standard l2 regularization is on the weights)
* weight-drop regularization - apply dropout to the weight matrices instead of activations
* learning rate scheduling - decrease the learning rate during training
* embedding dropout - zero out the entire embedding for a random set of words in the embedding matrix
* ensembling - average the predictions of several models trained with different initialization random seeds
* temporal activation regularization - add l2 regularization on the difference between the LSTM output activations at adjacent timesteps

You may notice that most of these suggestions are regularization techniques.  This dataset is considered fairly small, so regularization is one of the best ways to improve performance.


Run the cell below in order to train your improved LSTM and evaluate it.  

In [ ]:
## Feel free to copy your original LSTM solution down here to modify for your report if you'd like. 
# YOUR CODE [optionally] HERE
##
class LSTMNetwork(nn.Module):
    # a PyTorch Module that holds the neural network for your model

    def __init__(self):
        super().__init__()

        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.drop=nn.Dropout(p=0.5)
        self.drop_e=nn.Dropout(p=0.2)
        self.embedding=nn.Embedding(vocab_size,128)
        self.lstm = nn.LSTM(128,512,num_layers=3,dropout=0.5)
        self.linear_0 = nn.Linear(512,256)
        self.linear_1 = nn.Linear(256,128)
        self.final = nn.Linear(128,vocab_size)
        # END SOLUTION

    def forward(self, x, state):
        """Compute the output of the network.
        
        Note: In the Pytorch LSTM tutorial, the state variable is named "hidden":
        https://pytorch.org/tutorials/beginner/nlp/sequence_models_tutorial.html

        The torch.nn.LSTM documentation is quite helpful:
        https://pytorch.org/docs/stable/nn.html#lstm
    
        x - a tensor of int64 inputs with shape (seq_len, batch)
        state - a tuple of two tensors with shape (num_layers, batch, hidden_size)
                representing the hidden state and cell state of the of the LSTM.
        returns a tuple with two elements:
          - a tensor of log probabilities with shape (seq_len, batch, vocab_size)
          - a state tuple returned by applying the LSTM.
        """

        # Note that the nn.LSTM module expects inputs with the sequence 
        # dimension before the batch by default.
        # In this case the dimensions are already in the right order, 
        # but watch out for this since sometimes people put the batch first

        # YOUR CODE HERE
        embeddings=self.embedding(x)
        embeddings=self.drop_e(embeddings)
        lstm_out,lstm_state=self.lstm(embeddings,state)
        linear_out=self.linear_1(self.linear_0(self.drop(lstm_out)))
        final_out=self.final(linear_out)
        self.embedding.weight = self.final.weight
        return (final_out,lstm_state)
        # BEGIN SOLUTION
        # END SOLUTION

class LSTMModel:
    "A class that wraps LSTMNetwork to handle training and evaluation."

    def __init__(self):
        self.network = LSTMNetwork().cuda()

    def train(self,n_epochs=5):
        train_iterator = torchtext.data.BPTTIterator(train_dataset, batch_size=64, 
                                                     bptt_len=32, device='cuda')
        # Iterate over train_iterator with a for loop to get batches
        # each batch object has a .text and .target attribute with
        # token id tensors for the input and output respectively.

        # The initial state passed into the LSTM should be set to zero.

        #dataset = NeuralNgramDataset(ids(train_text), self.n)
        # train = torch.utils.data.DataLoader(dataset, batch_size=128, shuffle=True)
        # iterating over train_loader with a for loop will return a 2-tuple of batched tensors
        # the first tensor will be previous token ids with size (batch, n-1),
        # and the second will be the current token id with size (batch, )
        # you will need to move these tensors to GPU, e.g. by using the Tensor.cuda() function.

        # this will take some time to run; use tqdm.tqdm_notebook to get a progress bar 
        # (see Project 0 for example)

        # YOUR CODE HERE

        # BEGIN SOLUTION
        adam=torch.optim.Adam(self.network.parameters(),lr=0.00085)
        #scheduler = torch.optim.lr_scheduler.ExponentialLR(adam, gamma=0.1)

        for epoch in range(n_epochs):
          print("Epoch: ",epoch)
          state = (torch.autograd.variable(torch.zeros(3,64,512)).cuda(),torch.autograd.variable(torch.zeros(3,64,512)).cuda())
          #print("state")
          for batch in tqdm.tqdm_notebook(train_iterator, leave=False):
            words_batch, next_batch = batch.text.cuda(), batch.target.cuda()
            adam.zero_grad()
            out,state = self.network(words_batch,(state[0].detach(),state[1].detach()))
            loss = F.cross_entropy(torch.permute(out,(1,2,0)).cuda(), next_batch.permute(1,0).cuda())
            loss.backward()
            adam.step()
          #scheduler.step()
          perplexity = self.dataset_perplexity(validation_dataset)
          print('Perplexity: ',perplexity)
          previous_best_score = 35000
          if perplexity < previous_best_score:
            previous_best_score = perplexity 
            torch.save(self.network.state_dict(),'network.pt')

          self.network.load_state_dict(torch.load('network.pt'))

        # END SOLUTION

    def next_word_probabilities(self, text_prefix):
        "Return a list of probabilities for each word in the vocabulary."

        prefix_token_tensor = torch.tensor(ids(text_prefix), device='cuda').view(-1, 1)
        
        # YOUR CODE HERE

        # BEGIN SOLUTION
        self.network.eval()
        state = (torch.autograd.variable(torch.zeros(3,1,512)).cuda(),torch.autograd.variable(torch.zeros(3,1,512)).cuda())
        out,state=self.network(prefix_token_tensor,(state[0].detach(),state[1].detach()))
        softmax=(F.softmax(out,dim=2))[-1,:,:].squeeze()
        return(softmax.tolist())
        # END SOLUTION

    def dataset_perplexity(self, torchtext_dataset):
        "Return perplexity as a float."
        # Your code should be very similar to next_word_probabilities, but
        # run in a loop over batches. Use torch.no_grad() for extra speed.

        iterator = torchtext.data.BPTTIterator(torchtext_dataset, batch_size=64, bptt_len=32, device='cuda')

        # YOUR CODE HERE

        # BEGIN SOLUTION

        self.network.eval()
        log_probabilities=[]
        state = (torch.autograd.variable(torch.zeros(3,64,512)).cuda(),torch.autograd.variable(torch.zeros(3,64,512)).cuda())

        for batch in tqdm.tqdm_notebook(iterator, leave=False):
          with torch.no_grad():
            words_batch, next_batch = batch.text.cuda(),batch.target.cuda()
            out,state=self.network(words_batch,(state[0].detach(),state[1].detach()))
            softmax=F.softmax(out,dim=2)
            probs=torch.gather(softmax, 2, next_batch.unsqueeze(-1))
            for each in probs.squeeze(-1):
               for sent in each:
                log_probabilities.append(math.log(sent.item(),2))
        self.network.train()
        return 2**-np.mean(log_probabilities) 

        # END SOLUTION

lstm_model = LSTMModel()
lstm_model.train(30)

print('lstm validation perplexity:', lstm_model.dataset_perplexity(validation_dataset))
save_truncated_distribution(lstm_model, 'lstm_predictions.npy', short=False)


### Gradescope

To download this notebook, go to `File->Download .ipynb`.  Please rename the file to match the name in our file list.  You can download other outputs by clicking the > arrow near the top left and finding it under `Files`. When submitting your ipython notebooks, make sure everything runs correctly if the cells are executed in order starting from a fresh session.  Note that just because a cell runs in your current session doesn't mean it doesn't rely on code that you have already changed or deleted.  If the code doesn't take too long to run, we recommend re-running everything with `Runtime->Restart and run all...`.

Be sure to check the output of the autograder after it runs.  It should confirm that no files are missing and that the output files have the correct format.  Note that the test set perplexities shown by the autograder are on a completely different scale from your validation set perplexities due to truncating the distribution and selecting different text.  Don't worry if the values seem much worse.